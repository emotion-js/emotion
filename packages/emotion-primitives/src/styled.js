// @flow
import * as React from 'react'
import { View, Text, Image, StyleSheet } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign
} from './utils'
import { getStyles } from './getStyles'
import { convertToRNStyles } from './convertToRNStyles'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

// Evaluate the styles and convert them to React Native using styled component context
function evalStyles(context, Comp, styles) {
  // Convert styles with or without interpolations to React Native (StyleSheet.create)
  let otherStyles = convertToRNStyles.call(context, styles)

  return getStyles.call(context, otherStyles, context.props)
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
export function createStyled(component: React.ElementType) {
  let isPrimitive = isPrimitiveComponent(component)
  let pick = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...styles: *) {
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })

        const emotionStyles = evalStyles(this, Styled, styles)

        return React.createElement(
          component,
          pickAssign(pick, {}, props, {
            ref: props.innerRef,
            style: StyleSheet.flatten(emotionStyles)
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
