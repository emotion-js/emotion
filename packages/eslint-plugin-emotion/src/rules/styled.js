module.exports = {
  meta: {
    fixable: 'code'
  },
  create(context) {
    return {
      ImportDeclaration(node) {
        if (
          node.source.value === 'react-emotion' ||
          node.source.value === 'preact-emotion'
        ) {
          let newImportPath =
            node.source.value === 'react-emotion'
              ? '@emotion/styled'
              : '@emotion/preact-styled'
          let quote = node.source.raw.charAt(0)
          context.report({
            node: node.source,
            message: `styled should be imported from ${newImportPath}`,
            fix:
              node.specifiers.length === 1 &&
              node.specifiers[0].type === 'ImportDefaultSpecifier'
                ? fixer => {
                    return fixer.replaceText(
                      node.source,
                      quote + newImportPath + quote
                    )
                  }
                : undefined
          })
        }
      }
    }
  }
}
