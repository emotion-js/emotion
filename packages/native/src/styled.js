import { StyleSheet } from 'react-native'
import { createStyled } from '@emotion/primitives-core'

/**
 * a function that returns a styled component which render styles in React Native
 */
let styled = createStyled(StyleSheet)

export { styled }
