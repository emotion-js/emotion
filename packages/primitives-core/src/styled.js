// @flow
import * as React from 'react'
import { testAlwaysTrue, pickAssign, interleave } from './utils'
import { ThemeContext } from '@emotion/core'
import { createCss } from './css'

let defaultPickTest = prop => prop !== 'theme' && prop !== 'innerRef'

type options = {
  getShouldForwardProp: (cmp: React.ElementType) => (prop: string) => boolean
}

export function createStyled(
  StyleSheet: Object,
  { getShouldForwardProp = () => defaultPickTest }: options = {}
) {
  const css = createCss(StyleSheet)

  return function createEmotion(component: React.ElementType) {
    let pickTest = getShouldForwardProp(component)

    return function createStyledComponent(...rawStyles: *) {
      let styles

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles)
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      // $FlowFixMe
      let Styled = React.forwardRef((props, ref) => {
        return (
          <ThemeContext.Consumer>
            {theme => {
              let mergedProps = pickAssign(testAlwaysTrue, {}, props, {
                theme: props.theme || theme
              })
              let stylesWithStyleProp = styles
              if (props.style) {
                stylesWithStyleProp = styles.concat(props.style)
              }
              const emotionStyles = css.apply(mergedProps, stylesWithStyleProp)

              if (process.env.NODE_ENV !== 'production' && props.innerRef) {
                console.error(
                  'innerRef is no longer supported, please use ref instead'
                )
              }

              return React.createElement(
                component,
                pickAssign(pickTest, {}, props, {
                  ref: ref,
                  style: emotionStyles
                })
              )
            }}
          </ThemeContext.Consumer>
        )
      })
      // $FlowFixMe
      Styled.withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      Styled.displayName = `emotion(${getDisplayName(component)})`

      return Styled
    }
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
