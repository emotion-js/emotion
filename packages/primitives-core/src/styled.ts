import * as React from 'react'
import { interleave } from './utils'
import { ThemeContext } from '@emotion/react'
import { createCss } from './css'
import { AbstractStyleSheet } from './types'

let testOmitPropsOnComponent = (prop: string) =>
  prop !== 'theme' && prop !== 'as'

interface CreateStyledOptions {
  getShouldForwardProp(cmp: React.ElementType): (prop: string) => boolean
}

interface StyledOptions {
  shouldForwardProp?(prop: string): boolean
}

type StyledProps = Record<string, unknown> & {
  as?: React.ElementType
}

export function createStyled(
  StyleSheet: AbstractStyleSheet,
  options?: CreateStyledOptions
) {
  const getShouldForwardProp =
    options?.getShouldForwardProp ?? (() => testOmitPropsOnComponent)

  const css = createCss(StyleSheet)

  return function createEmotion(
    component: React.ElementType,
    options?: StyledOptions
  ) {
    let shouldForwardProp =
      options && options.shouldForwardProp
        ? options.shouldForwardProp
        : undefined
    let defaultShouldForwardProp =
      shouldForwardProp || getShouldForwardProp(component)
    let shouldUseAs = !defaultShouldForwardProp('as')

    return function createStyledComponent(...rawStyles: any[]) {
      let styles: any[]

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles as [any, ...any[]])
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      let Styled = React.forwardRef<unknown, StyledProps>((props, ref) => {
        const finalTag = (shouldUseAs && props.as) || component

        let mergedProps = props
        if (props.theme == null) {
          mergedProps = {}
          for (let key in props) {
            mergedProps[key] = props[key]
          }
          mergedProps.theme = React.useContext(ThemeContext)
        }

        let finalShouldForwardProp =
          shouldUseAs && shouldForwardProp === undefined
            ? getShouldForwardProp(finalTag)
            : defaultShouldForwardProp

        let newProps: Record<string, unknown> = {}

        for (let key in props) {
          if (shouldUseAs && key === 'as') continue

          if (finalShouldForwardProp(key)) {
            newProps[key] = props[key]
          }
        }

        newProps.style = [css.apply(mergedProps, styles), props.style]
        newProps.ref = ref

        return React.createElement(finalTag, newProps)
      })

      Styled.displayName = `emotion(${getDisplayName(component)})`

      const withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      const castedStyled = Styled as typeof Styled & {
        withComponent: typeof withComponent
      }

      castedStyled.withComponent = withComponent

      return castedStyled
    }
  }
}

const getDisplayName = (
  primitive: string | { displayName?: string; name?: string }
) =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
