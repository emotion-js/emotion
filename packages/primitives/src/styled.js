// @flow
import * as React from 'react'
import { View, Text, Image } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign,
  interleave
} from './utils'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'
import { css } from './css'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */

type CreateStyledComponent = (...styles: any) => React.ElementType

type BaseCreateStyled = (tag: React.ElementType) => CreateStyledComponent

export type CreateStyled = BaseCreateStyled & {
  View: CreateStyledComponent,
  Text: CreateStyledComponent,
  Image: CreateStyledComponent
}

// $FlowFixMe
let createStyled: CreateStyled = function createStyled(
  component: React.ElementType
) {
  let isPrimitive = isPrimitiveComponent(component)
  let pickTest = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...rawStyles: *) {
    let styles

    if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
      styles = rawStyles
    } else {
      styles = interleave(rawStyles)
    }
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })
        let stylesWithStyleProp = styles
        if (props.style) {
          stylesWithStyleProp = styles.concat(props.style)
        }
        const emotionStyles = css.apply(this, stylesWithStyleProp)

        return React.createElement(
          component,
          pickAssign(pickTest, {}, props, {
            ref: props.innerRef,
            style: emotionStyles
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'

export { createStyled }
