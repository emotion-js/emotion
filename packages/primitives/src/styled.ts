import * as React from 'react'
import { StyleSheet, View, Text, Image } from 'react-primitives'
import { createStyled } from '@emotion/primitives-core'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function getShouldForwardProp(
  component: React.ElementType
): (prop: string) => boolean {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return testPickPropsOnPrimitiveComponent
    }
  }
  return testPickPropsOnOtherComponent
}

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
type CreateStyledComponent = (...styles: any) => React.FC<any> & {
  withComponent: (component: any) => React.FC<any>
}

/** @internal */
export type BaseStyled = (tag: React.ElementType) => CreateStyledComponent

export type Styled = BaseStyled & {
  View: CreateStyledComponent
  Text: CreateStyledComponent
  Image: CreateStyledComponent
}

let styled: BaseStyled = createStyled(StyleSheet, { getShouldForwardProp })

export { styled }
