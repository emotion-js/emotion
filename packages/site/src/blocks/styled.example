const Image = ({ className, src }) => <img src={src} className={className} />

const Avatar = styled(Image)`
  width: 96px;
  height: 96px;
  border-radius: 50%;
  transition: transform 400ms ease-in-out;

  &:hover {
    transform: scale(1.2);
  }
`

render(<Avatar src={logoUrl} />)
