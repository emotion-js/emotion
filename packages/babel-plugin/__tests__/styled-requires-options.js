// @flow
import babelTester from 'babel-tester'
import plugin from '@emotion/babel-plugin'

const cases = {
  'hash generation no file system': {
    code: `
import styled from '@emotion/styled'
styled.h1\`color:blue;\`
`,
    plugins: [plugin],
    babelFileName: undefined
  },
  'autoLabel set to never': {
    code: `
import styled from '@emotion/styled'
const Foo = styled.h1\`color:hotpink;\`
    `,
    plugins: [[plugin, { autoLabel: 'never' }]],
    babelFileName: __filename
  }
}

babelTester('styled inline', cases)
