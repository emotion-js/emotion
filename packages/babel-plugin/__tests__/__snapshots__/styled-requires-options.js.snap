// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled inline autoLabel set to never 1`] = `
"
import styled from '@emotion/styled'
const Foo = styled.h1\`color:hotpink;\`
    

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"@emotion/styled/base\\";

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return \\"You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop).\\"; }

const Foo = _styled(\\"h1\\", {
  target: \\"eoxfmb00\\"
})(process.env.NODE_ENV === \\"production\\" ? {
  name: \\"3sn2xs\\",
  styles: \\"color:hotpink\\"
} : {
  name: \\"3sn2xs\\",
  styles: \\"color:hotpink\\",
  map: \\"/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlZC1yZXF1aXJlcy1vcHRpb25zLmpzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUVxQiIsImZpbGUiOiJzdHlsZWQtcmVxdWlyZXMtb3B0aW9ucy5qcyIsInNvdXJjZXNDb250ZW50IjpbIlxuaW1wb3J0IHN0eWxlZCBmcm9tICdAZW1vdGlvbi9zdHlsZWQnXG5jb25zdCBGb28gPSBzdHlsZWQuaDFgY29sb3I6aG90cGluaztgXG4gICAgIl19 */\\",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
});"
`;

exports[`styled inline hash generation no file system 1`] = `
"
import styled from '@emotion/styled'
styled.h1\`color:blue;\`


      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from \\"@emotion/styled/base\\";

_styled(\\"h1\\", {
  target: \\"e8evh370\\"
})({
  name: \\"117wnve\\",
  styles: \\"color:blue\\"
});"
`;
